src/machine.rs

```diff
@@ -280,6 +287,9 @@ impl<'mir, 'tcx> Machine<'mir, 'tcx> for Evaluator<'tcx> {
                 let data = vec![0; size.bytes() as usize];
                 Allocation::from_bytes(&data, tcx.data_layout.pointer_align.abi)
             }
+            "environ" => {
+                memory_extra.environ.as_ref().unwrap().clone()
```

Comment:
Hmm but then what should happen if you set another variable? I haven't thought about that part yet

Reply:
^^ that is handled by miri. You only need to provide the initial value here, afterwards you have a perfectly viable mutable allocation.

Reply:
what do you mean? shouldn't we need to update `environ` every time `(un)setenv` is called?

Reply:
I gotta run, so just an excercise and no explanation: go up the call stack from here (you will have to go up a few call frames into the miri engine in rustc) and check out what happens with your allocation and why this code is even triggered.

Reply:
Also see https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir/interpret/trait.Machine.html#tymethod.find_foreign_static: this function is only called the *first* time a foreign static is accessed.

@oli-obk I wonder... would it make sense to expand/change the `find_foreign_static` type so that it can/must return an `AllocId`? Having to actually return the *allocation* here is a bit annoying for `environ`.

Reply:
That would mean we'd end up with two `AllocId`s for the same allocation. I guess that happens anyway for immutable statics and the initial value of mutable statics, unsure. Maybe we should rather pass *in* the `AllocId` and require `find_foreign_static` to put the allocation into `Memory` at that alloc id. We'd need an `allocate_with(AllocId, ...)` function, but then everything else could just work as intended without creating confusing situations.

Reply:
Hm... we could pull the same trick as with "normal" statics, and treat one `AllocId` as "hidden"? We have two `AllocId` there and it's working...

I don't think `allocate_with(alloc_id)` will work; we cannot change what that ID maps to in the global tcx store as that's immutable -- right?

Reply:
All non interned allocations need to be in the memory's alloc_map, which is the first thing that is always queried. If there's an entry there, no further code is looked at. Theoretically we could proactively insert that allocation at memory creation time instead of the current "lazy" (it's just lazily inserting, the creation is eager) scheme.

Reply:
The main problem with this (using `Option::take`) is that we would need `&mut MemoryExtra` as a parameter and that would require changing a lot of code. Maybe we can add interior mutability to that field and call it for a day?